File: portfolio_flutter/lib/MovieRepository.dart

import 'package:http/http.dart' as http;
import 'dart:convert';

class MovieRepository {
  Future<List<MovieItem>?> fetchMovies(int page) async {
    final apiUrl =
        "https://api.themoviedb.org/3/movie/popular?api_key=a38bb7041808747f410e16a68a4ac2ae&page=$page";

    try {
      final response = await http.get(Uri.parse(apiUrl));
      if (response.statusCode == 200) {
        final Map<String, dynamic> data = json.decode(response.body);
        final List<dynamic> results = data['results'];

        List<MovieItem> movies = results
            .map((movieData) => MovieItem.fromJson(movieData))
            .toList();

        return movies;
      } else {
        print("Erro na solicitação: ${response.reasonPhrase}");
        return null;
      }
    } catch (error) {
      print("Erro na solicitação: $error");
      return null;
    }
  }
}

class MovieResponse {
  late List<MovieItem>? results;

  MovieResponse({required this.results});

  MovieResponse.fromJson(Map<String, dynamic> json) {
    if (json['results'] != null) {
      results = <MovieItem>[];
      json['results'].forEach((v) {
        results!.add(MovieItem.fromJson(v));
      });
    }
  }
}

class MovieItem {
  late String id;
  late String title;
  late String overview;
  late String posterPath;

  MovieItem({
    required this.id,
    required this.title,
    required this.overview,
    required this.posterPath,
  });

  MovieItem.fromJson(Map<String, dynamic> json) {
    id = json['id'].toString();
    title = json['title'];
    overview = json['overview'];
    posterPath = json['poster_path'];
  }
}

File: iosMovie/firstProject/MovieRepository.swift
//
//  MovieRepository.swift
//  firstProject
//
//  Created by Caio Henrique on 12/10/23.
//

import Foundation
import Alamofire

class MovieRepository {
    func fetchMovies(completion: @escaping ([MovieItem]?) -> Void) {
        let apiUrl = "https://api.themoviedb.org/3/movie/popular?api_key=a38bb7041808747f410e16a68a4ac2ae"
        
        AF.request(apiUrl).responseDecodable(of: MovieResponse.self) { response in
            switch response.result {
            case .success(let results):
                completion(results.results)
            case .failure(let error):
                print("Erro na solicitação: \(error)")
                completion(nil)
            }
        }
    }
}


struct MovieResponse: Identifiable, Decodable {
    let id = UUID()
    let results: [MovieItem]?

    private enum CodingKeys: String, CodingKey {
        case results
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        results = try container.decode([MovieItem].self, forKey: .results)
    }
}

struct MovieItem: Identifiable, Decodable {
    let id = UUID()
    let title: String
    let overview: String
    let poster_path: String


    private enum CodingKeys: String, CodingKey {
        case title
        case overview
        case poster_path
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        title = try container.decode(String.self, forKey: .title)
        overview = try container.decode(String.self, forKey: .overview)
        poster_path = try container.decode(String.self, forKey: .poster_path)
    }
}
